tests/host_tools/uffd/src/bin/valid_handler.rs

```diff
@@ -0,0 +1,289 @@
+    backing_buffer: *const u8,
+    uffd: Uffd,
+    // Not currently used but included to demonstrate how a page fault handler can
+    // fetch Firecracker's PID in order to make it aware of any crashes/exits.
+    _firecracker_pid: u32,
+}
+
+#[derive(Clone)]
+enum MemPageState {
+    Uninitialized,
+    FromFile,
+    Removed,
+    Anonymous,
+}
+
+impl UffdPfHandler {
+    pub fn from_unix_stream(stream: UnixStream, data: *const u8, size: usize) -> Self {
+        let mut message_buf = vec![0u8; 1024];
+        let (bytes_read, file) = stream
+            .recv_with_fd(&mut message_buf[..])
+            .expect("Cannot recv_with_fd");
+        message_buf.resize(bytes_read, 0);
+
+        let body = String::from_utf8(message_buf).unwrap();
+        let file = file.expect("Uffd not passed through UDS!");
+
+        let mappings = serde_json::from_str::<Vec<GuestRegionUffdMapping>>(&body)
+            .expect("Cannot deserialize memory mappings.");
+        let memsize: usize = mappings.iter().map(|r| r.size).sum();
+
+        // Make sure memory size matches backing data size.
+        assert_eq!(memsize, size);
+
+        let uffd = unsafe { Uffd::from_raw_fd(file.into_raw_fd()) };
+
+        let creds: libc::ucred = get_peer_process_credentials(stream);
+
+        let mem_regions = create_mem_regions(&mappings);
+
+        Self {
+            mem_regions,
+            backing_buffer: data,
+            uffd,
+            _firecracker_pid: creds.pid as u32,
+        }
+    }
+
+    fn update_mem_state_mappings(&mut self, start: u64, end: u64, state: &MemPageState) {
+        for region in self.mem_regions.iter_mut() {
+            for (key, value) in region.page_states.iter_mut() {
+                if key >= &start && key < &end {
+                    *value = state.clone();
+                }
+            }
+        }
+    }
+
+    fn populate_from_file(&self, region: &MemRegion) -> (u64, u64) {
+        let src = self.backing_buffer as u64 + region.mapping.offset;
+        let start_addr = region.mapping.base_host_virt_addr;
+        let len = region.mapping.size;
+        // Populate whole region from backing mem-file.
+        // This offers an example of how memory can be loaded in RAM,
+        // however this can be adjusted to accommodate use case needs.
+        let ret = unsafe {
+            self.uffd
+                .copy(src as *const _, start_addr as *mut _, len, true)
+                .expect("Uffd copy failed")
+        };
+
+        // Make sure the UFFD copied some bytes.
+        assert!(ret > 0);
+
+        return (start_addr, start_addr + len as u64);
+    }
+
+    fn zero_out(&mut self, addr: u64) -> (u64, u64) {
+        let page_size = get_page_size().unwrap();
+
+        let ret = unsafe {
+            self.uffd
+                .zeropage(addr as *mut _, page_size, true)
+                .expect("Uffd zeropage failed")
+        };
+        // Make sure the UFFD zeroed out some bytes.
+        assert!(ret > 0);
```

Comment:
what if it was already all zeroed out?

Reply:
Looking at the [uffd kernel code](https://elixir.bootlin.com/linux/v4.8/source/fs/userfaultfd.c#L1101), the kernel does not care for the contents of the page before zeroing. The `zeropage()` function returns 1 if operation succeeded (`pagesize` zeros were written at the specified address) and -1 in case an error occurs.